homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,195 @@
+"""
+Sensor to collect the reference daily prices of electricity ('PVPC') in Spain.
+
+For more details about this platform, please refer to the documentation at
+https://www.home-assistant.io/integrations/pvpc_hourly_pricing/
+"""
+from datetime import timedelta
+import logging
+from random import randint
+from typing import Optional
+
+from aiopvpc import PVPCData
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.event import (
+    async_track_point_in_time,
+    async_track_time_change,
+)
+from homeassistant.helpers.restore_state import RestoreEntity
+import homeassistant.util.dt as dt_util
+
+from .const import ATTR_TARIFF
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_PRICE = "price"
+ICON = "mdi:currency-eur"
+UNIT = "€/kWh"
+
+_DEFAULT_TIMEOUT = 10
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: config_entries.ConfigEntry, async_add_entities
+):
+    """Set up the electricity price sensor from config_entry."""
+    name = config_entry.data[CONF_NAME]
+    pvpc_data_handler = PVPCData(
+        tariff=config_entry.data[ATTR_TARIFF],
+        local_timezone=hass.config.time_zone,
+        websession=async_get_clientsession(hass),
+        logger=_LOGGER,
+        timeout=_DEFAULT_TIMEOUT,
+    )
+    async_add_entities(
+        [ElecPriceSensor(name, config_entry.unique_id, pvpc_data_handler)], True
+    )
+
+
+class ElecPriceSensor(RestoreEntity):
+    """Class to hold the prices of electricity as a sensor."""
+
+    unit_of_measurement = UNIT
+    icon = ICON
+    should_poll = False
+
+    def __init__(self, name, unique_id, pvpc_data_handler):
+        """Initialize the sensor object."""
+        self._name = name
+        self._unique_id = unique_id
+        self._pvpc_data = pvpc_data_handler
+        self._num_retries = 0
+
+        self._init_done = False
+        self._hourly_tracker = None
+        self._price_tracker = None
+
+    async def async_will_remove_from_hass(self) -> None:
+        """Cancel listeners for sensor updates."""
+        self._hourly_tracker()
+        self._price_tracker()
+
+    async def async_added_to_hass(self):
+        """Handle entity which will be added."""
+        await super().async_added_to_hass()
+        state = await self.async_get_last_state()
+        if state:
+            self._pvpc_data.state = state.state
+
+        # Update 'state' value in hour changes
+        self._hourly_tracker = async_track_time_change(
```

Comment:
If the user removes the integration, these will still be running in the background but the integration will be unloaded.

Since you are storing them you could cancel them on removal:
```
async def async_will_remove_from_hass(self):
```

Reply:
Is already there :)

```
    async def async_will_remove_from_hass(self) -> None:
        """Cancel listeners for sensor updates."""
        self._hourly_tracker()
        self._price_tracker()
```

Reply:
Yes it is . Sorry I missed that!

Reply:
> Sorry I missed that!

by no means, I'm enjoying a lot your exhaustive review, please continue :)